// trulioo-react/EmbedID.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = EmbedID;

var _react = _interopRequireDefault(require("react"));

var _reactRedux = require("react-redux");

var _redux = require("redux");

var _reduxThunk = _interopRequireDefault(require("redux-thunk"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reducers = _interopRequireDefault(require("./reducers"));

var _TruliooForm = _interopRequireDefault(require("./components/TruliooForm"));

var store = (0, _redux.createStore)(_reducers.default, (0, _redux.applyMiddleware)(_reduxThunk.default));

function EmbedID(_ref) {
  var handleResponse = _ref.handleResponse,
      url = _ref.url,
      handleSubmit = _ref.handleSubmit,
      customFields = _ref.customFields;
  return _react.default.createElement(_reactRedux.Provider, {
    store: store
  }, _react.default.createElement(_TruliooForm.default, {
    handleResponse: handleResponse,
    url: url,
    handleSubmit: handleSubmit,
    customFields: customFields
  }));
}

EmbedID.propTypes = {
  handleResponse: _propTypes.default.func,
  handleSubmit: _propTypes.default.func,
  url: _propTypes.default.string,
  customFields: _propTypes.default.objectOf(_propTypes.default.object)
};
EmbedID.defaultProps = {
  handleResponse: false,
  url: true,
  handleSubmit: undefined,
  customFields: undefined
};
